### PR TITLE
#164047060 Password reset request through email

### PR DESCRIPTION
**Pull request by kevpy**
#### Description
___
When a user requests for a password reset, they should be able to receive a email with a link to reset password. A JWT token is sent with the link and used later for verification

#### Type of Change
___

- [x] New Feature

#### How has this been tested
___
- [x] Unit tests
 - [x]  End to end testing

#### Checklist
___
- [x] Tests are included to show/proof the feature works.

#### Pivotal Tracker
___
[164047060](https://www.pivotaltracker.com/story/show/164047060)

### CODE REVIEW

File: authors/apps/authentication/models.py [hunk 1]
```diff
@@ -126,16 +128,121 @@ def get_short_name(self):
...
+        :param tkn:
+        :return: none
+        """
+        reset_tkn = PasswordTokenReset.objects.create(user=user, token=tkn)
+        reset_tkn.save()
+        # return reset_tkn
+
+    @staticmethod
+    def generate_link_message(token):
+        """Tjis method takes in a token and generates a link to be
```
Comment (JoanNgatia) [inline authors/apps/authentication/models.py]: Fix typo error `tjis` to `this`

File: authors/apps/authentication/models.py [hunk 2]
```diff
@@ -126,16 +128,121 @@ def get_short_name(self):
...
+                          "for Authors Haven has requested a password reset." \
+                          " If you did, click on the given link" \
+                          " \n {} \notherwise ignore this email. ".format(link)
+        return request_message
+
+    @staticmethod
+    def save_new_password(user_credentials, password):
+        """This method takes in a token verifies it and resets users password
+        :param reset_token: token sent with a reset link
+        :return:
```
Comment (JoanNgatia) [inline authors/apps/authentication/models.py]: State that the return is a message showing success

File: authors/apps/authentication/models.py [hunk 3]
```diff
@@ -126,16 +128,121 @@ def get_short_name(self):
...
         Generates a JSON Web Token that stores this user's ID and has an expiry
         date set to 60 days into the future.
         """
-        dt = datetime.now() + timedelta(days=60)
+        dt = datetime.now() + timedelta(days)
 
         token = jwt.encode({
             'id': self.pk,
             'exp': int(dt.strftime('%s'))
         }, settings.SECRET_KEY, algorithm='HS256')
```
Comment (JoanNgatia) [inline authors/apps/authentication/models.py]: Good work on the token hashing! 🔥

File: authors/apps/authentication/models.py [hunk 4]
```diff
@@ -126,16 +128,121 @@ def get_short_name(self):
...
         return self.username
 
-    def _generate_jwt_token(self):
+    def _generate_jwt_token(self, days):
         """
         Generates a JSON Web Token that stores this user's ID and has an expiry
         date set to 60 days into the future.
         """
-        dt = datetime.now() + timedelta(days=60)
+        dt = datetime.now() + timedelta(days)
```
Comment (JoanNgatia) [inline authors/apps/authentication/models.py:137]: Can we leave this with the `60 days` constraint?

**Review by JoanNgatia — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/models.py (hunk 4 above)
Comment (kevpy) [reply authors/apps/authentication/models.py:137]: The token hashing is used by two features. One for authentication is set to 60 days whereas the one for password reset I had to limit it to a day. Thought it wise that a reset token shouldn't last as long. Happy to be advised differently though.

**Comment by Kibetchirchir** [comment]
Nice work. I like how you handled the token expiration 👍

File: authors/apps/authentication/models.py (hunk 4 above)
Comment (JoanNgatia) [reply authors/apps/authentication/models.py:137]: This approach is okay and yes, it is better for a reset token to not last as long.

**Review by JoanNgatia — APPROVED** [review]

**Pull request merged**